### PR TITLE
[WIP] Observable Tree

### PR DESCRIPTION
**Pull request by srdanrasic**
@tonyarnold I've merged your PR into `observable-tree` branch so we can work here. 

I've also refactored things a bit, but I've just started. We don't need (Mutable)ObservableTreeNode classes so I removed them. We can use `ObservableCollection` where underlying collection is  `TreeNode`. I've renamed `TreeNode` into `TreeNodeProtocol` and introduced a concrete tree implementation - `TreeNode`.

We need to add more methods to `TreeNodeProtocol` and then add respective variants to `ObservableCollection+TreeNode`.

I also need to handle batched diff and patching properly. 

Check the playground.

### CODE REVIEW

File: Sources/Bond/ObservableTree/TreeNodeProtocol.swift [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+public protocol TreeNodeProtocol: MutableCollection where Index == IndexPath, Element == Self {
+
+    associatedtype Value
+    var value: Value { get }
+    var children: [Self] { get }
```
Comment (tonyarnold) [inline Sources/Bond/ObservableTree/TreeNodeProtocol.swift]: Is there a way to express this without forcing it to be an array? Making this Collection or MutableCollection would solve some headaches for me.

Comment (tonyarnold) [reply Sources/Bond/ObservableTree/TreeNodeProtocol.swift]: It'd also be nice to have a "root" tree node that returns `value` as `nil`, or a type that's not the same as `Element`. 

I have a structure I'm working against that is:

- Page
    - Layer
    - Layer
    - Layer
        - Layer
        - Layer
    - Layer

etc, etc.

Comment (srdanrasic) [reply Sources/Bond/ObservableTree/TreeNodeProtocol.swift]: Yes, I think we can make children a collection.

As for the `Value` type (not to be confused with `Element` which is another `ThreeNode<Value>`), could you maybe define your tree as `TreeNode<Layer?>`? It should have the same effect. 

Alternatively, maybe a better solution would be to use enum to define all possible tree node type:

```
enum Item {
    case page(Page)
    case layer(Layer)
}
```

Comment (tonyarnold) [reply Sources/Bond/ObservableTree/TreeNodeProtocol.swift]: How do you see the enum work in this context? That's obviously preferable, but I'm not sure how it fits alongside what we're designing here.

Comment (srdanrasic) [reply Sources/Bond/ObservableTree/TreeNodeProtocol.swift]: Would that work for you? I don't think it's possible to make a tree with different value types on different levels, at least no without having an infinite number of generic arguments 😅

**Comment by tonyarnold** [comment]
So bear with me here, I want to confirm a couple of basics in the use of this feature:

- Is the intent that all edits to the tree should be done via the "root" node, with an appropriate `IndexPath`? 
- If so, does this mean that we just need to observe the root node for diffs/patches. 

That's what I was aiming for with my initial work - I just want to see if you're thinking the same. 

I've implemented a TreeNodeProtocol conformer in my local project, and that seems to match what I was expecting, so it's all looking very promising!

**Comment by tonyarnold** [comment]
@srdanrasic what was the driver behind using `RangeReplaceableCollection` to implement editing here? I am having hell's own trouble retrofitting some Objective-C `NSFastEnumeration` classes to implement `RangeReplaceableCollection` - is there an alternative way to come at editing?

**Comment by srdanrasic** [comment]
Hey @tonyarnold, how would you implement inserting or deleting objects with those classes? NSFastEnumeration is just an Objective-C style Sequence. So it does not, by itself, support any mutation. 

Btw, I'm using our own `RangeReplacableTreeNode` to implement editing, not `RangeReplaceableCollection`. We can change the protocol if needed. The idea is that the conforming class has to implement only one method, `replaceSubrange`. It then gets all other editing methods for free.

**Comment by tonyarnold** [comment]
I am in an odd situation where I have an Objective-C wrapper class sitting around a C++ class that does the heavy lifting of insert/update/delete/etc. The class presents with an API that’s almost identical to `NSArray`/`NSMutableArray`, but isn’t a subclass of either. This is why I’ve been pushing pretty hard on using Collection over Array - I can make these custom classes present with most of the Collection conformances (although `RangeReplaceableCollection` eludes me at this point).

The implementation of `RangeReplacableTreeNode `is calling `replaceSubrange()` on the collection, so I assumed that it needed the collection to conform to `RangeReplaceableCollection`.

**Comment by srdanrasic** [comment]
Good point, maybe we should rename that to `replaceChildrenSubrange` or something like that. Given that you don't have to conform to `RangeReplaceableCollection`, would you be able to implement `replaceChildrenSubrange` using the wrapper's existing API?

**Comment by tonyarnold** [comment]
Potentially, yes - I'm talking to the C++ developer on the project about providing an analogue to `NSMutableArray`'s API for this:

```swift
func replaceObjects(in range: NSRange, withObjectsFrom otherArray: [Any])
```

As always, you do what you think it is the best outcome for Bond. I'm basically mirroring the editing APIs in my project for these custom list classes so that I can workaround differences in indexing, so I can just update my local versions when the core versions change.

**Comment by srdanrasic** [comment]
Alright, that method seems like the thing you are missing. I'm sure it can be implemented with a right combination of `insert(at:)` and `remove(at:)` methods. 

My goal for Bond is to support something like that 😄 

Note that batching, patching and diffing of the trees is not yet implemented. I'll be working on that today and over the weekend.

**Comment by srdanrasic** [comment]
Batching, patching and diffing of the trees is implemented 🎉  

```swift
tree.batchUpdate { ... }
tree.replace(with: otherTree, performDiff: true)
```

There might be some bugs so I'll write some unit tests during the weekend. In the meantime, I'm merging this PR so we can start finalizing the huge work on observable collections 😅

**Pull request merged**